invokeai/frontend/web/src/app/components/App.tsx

```diff
@@ -70,6 +73,19 @@ const App = ({ config = DEFAULT_CONFIG, selectedImage, destination }: Props) =>
     }
   }, [dispatch, config, logger]);
 
+  const { getAndLoadWorkflow } = useGetAndLoadLibraryWorkflow({
+    onSuccess: () => {
+      setActiveTab('workflows');
+    },
+  });
+
+  useEffect(() => {
+    if (selectedWorkflow && !hasLoadedRef.current) {
+      getAndLoadWorkflow(selectedWorkflow);
+      hasLoadedRef.current = true;
```

Comment:
Used a ref instead of state or getAndLoadWorkflowResult.isLoading since that was the only way I could get this to fire only once, open to other suggestions though!

Reply:
Yep! I tried doing it that way too and it still would rerender too many times and load the workflow a bunch

Reply:
I took a crack at it and do see it hitting that hook twice, but I think the app.tsx component itself loads multiple times so that would be expected. It deson't seem to interrupt the UX though so I think its fine. Are you just seeing the two times or more than that?

Reply:
I was seeing it three times and so it would display on the toast "Workflow Loaded (3)"

Reply:
It's expected for hooks to be called twice in dev mode when using the `StrictMode` component, which wraps the app.

In strict mode, all hooks are called twice initially as a way of forcing you to write them correctly. If your hook doesn't do what you think it should when called twice, it's probably doing it wrong. It's a particularly important check for hooks that should do some cleanup when the component unmounts.

In production, you'll only get one call. Remove the `StrictMode` component to see what the behaviour would be in the prod build.

There's another issue, though. The arg passed to `useGetAndLoadLibraryWorkflow` with the `onSuccess` callback is an object. That means when `App` re-renders for any reason, the `useGetAndLoadLibraryWorkflow` hook will be run again, causing unexpected additional workflow loads.

To fix this, we could memoize the options:
```tsx
  const loadWorkflowOptions = useMemo(
    () => ({
      onSuccess: () => {
        dispatch(setActiveTab('workflows'));
      },
    }),
    [dispatch]
  );

  const { getAndLoadWorkflow } = useGetAndLoadLibraryWorkflow(loadWorkflowOptions);
```

Then you don't need the ref. But, as described in the other comment, you don't need the `onSuccess` at all, because the tab is being set via an "extra" reducer.

---

The `useGetAndLoadLibraryWorkflow` hook wants an arg as an object, but really it should be OK without an arg. Let's update that hook like this, so no arg is required:

```diff
diff --git a/invokeai/frontend/web/src/features/workflowLibrary/hooks/useGetAndLoadLibraryWorkflow.ts b/invokeai/frontend/web/src/features/workflowLibrary/hooks/useGetAndLoadLibraryWorkflow.ts
index 89933999b..1888cbd54 100644
--- a/invokeai/frontend/web/src/features/workflowLibrary/hooks/useGetAndLoadLibraryWorkflow.ts
+++ b/invokeai/frontend/web/src/features/workflowLibrary/hooks/useGetAndLoadLibraryWorkflow.ts
@@ -15,9 +15,9 @@ type UseGetAndLoadLibraryWorkflowReturn = {
   getAndLoadWorkflowResult: ReturnType<typeof useLazyGetWorkflowQuery>[1];
 };
 
-type UseGetAndLoadLibraryWorkflow = (arg: UseGetAndLoadLibraryWorkflowOptions) => UseGetAndLoadLibraryWorkflowReturn;
+type UseGetAndLoadLibraryWorkflow = (arg?: UseGetAndLoadLibraryWorkflowOptions) => UseGetAndLoadLibraryWorkflowReturn;
 
-export const useGetAndLoadLibraryWorkflow: UseGetAndLoadLibraryWorkflow = ({ onSuccess, onError }) => {
+export const useGetAndLoadLibraryWorkflow: UseGetAndLoadLibraryWorkflow = (arg) => {
   const dispatch = useAppDispatch();
   const toast = useToast();
   const { t } = useTranslation();
@@ -29,17 +29,17 @@ export const useGetAndLoadLibraryWorkflow: UseGetAndLoadLibraryWorkflow = ({ onS
         // This action expects a stringified workflow, instead of updating the routes and services we will just stringify it here
         dispatch(workflowLoadRequested({ data: { workflow: JSON.stringify(workflow), graph: null }, asCopy: false }));
         // No toast - the listener for this action does that after the workflow is loaded
-        onSuccess && onSuccess();
+        arg?.onSuccess && arg.onSuccess();
       } catch {
         toast({
           id: `AUTH_ERROR_TOAST_${workflowsApi.endpoints.getWorkflow.name}`,
           title: t('toast.problemRetrievingWorkflow'),
           status: 'error',
         });
-        onError && onError();
+        arg?.onError && arg.onError();
       }
     },
-    [_getAndLoadWorkflow, dispatch, onSuccess, t, onError, toast]
+    [_getAndLoadWorkflow, dispatch, arg, toast, t]
   );
 
   return { getAndLoadWorkflow, getAndLoadWorkflowResult };
```

---

BTW - I think the ref would have an unintended consequence. If we had previously passed a `selectedWorkflowId` to `InvokeAIUI`, then change the `selectedWorkflowId` prop, the ref will prevent it from loading even though we are requesting a different workflow.

Perhaps a better way to handle that situation would have been to add some logic in `useGetAndLoadLibraryWorkflow` or the [`workflowLoadRequested listener`](https://github.com/invoke-ai/InvokeAI/blob/main/invokeai/frontend/web/src/app/store/middleware/listenerMiddleware/listeners/workflowLoadRequested.ts) to not load the same workflow twice.

But in this case, I think it's OK that we load the workflow twice in dev mode.